framework_tvb/tvb/core/services/backend_clients/hpc_scheduler_client.py

```diff
@@ -0,0 +1,425 @@
+                if not operation.has_finished:
+                    operation.mark_complete(STATUS_ERROR)
+                    dao.store_entity(operation)
+                return
+            h5_filenames = HPCSchedulerClient._stage_out_to_operation_folder(wd, operation)
+            message = HPCSchedulerClient()._update_db_with_results(operation, h5_filenames)
+            LOGGER.debug(message)
+        except HTTPError as exception:
+            LOGGER.error("Failed to submit job HPC")
+            operation.mark_complete(STATUS_ERROR, exception.response.text)
+            dao.store_entity(operation)
+
+    @staticmethod
+    def _stage_out_outputs(h5_path, output_list):
+        # type: (str, dict) -> list
+        result_filenames = []
+        for output_filename, output_filepath in output_list.items():
+            if type(output_filepath) is not unicore_client.PathFile:
+                continue
+            filename = os.path.join(h5_path, os.path.basename(output_filename))
+            output_filepath.download(filename)
+            result_filenames.append(filename)
+        return result_filenames
+
+    @staticmethod
+    def execute(operation_id, user_name_label, adapter_instance):
+        # type: (int, None, None) -> None
+        """Call the correct system command to submit a job to HPC."""
+        # HPCSchedulerClient._run_hpc_job(operation_id)
+
+        thread = HPCOperationThread(operation_id, target=HPCSchedulerClient._run_hpc_job,
+                                    kwargs={'operation_identifier': operation_id})
+        thread.start()
+        HPC_THREADS.append(thread)
+
+    @staticmethod
+    def stop_operation(operation_id):
+        # TODO: Review this implementation after DAINT maintenance
+        operation = dao.get_operation_by_id(operation_id)
+        if not operation or operation.has_finished:
+            LOGGER.warning("Operation already stopped: %s" % operation_id)
+            return True
+
+        LOGGER.debug("Stopping HPC operation: %s" % str(operation_id))
+        op_ident = OperationDAO().get_operation_process_for_operation(operation_id)
+        if op_ident is not None:
+            # TODO: Handle login
+            transport = unicore_client.Transport(os.environ[HPCSchedulerClient.CSCS_LOGIN_TOKEN_ENV_KEY])
+            # Abort HPC job
+            job = Job(transport, op_ident.job_id)
+            if job.is_running():
+                job.abort()
+
+        # Kill thread
+        operation_thread = get_op_thread(operation_id)
+        if operation_thread is None:
+            LOGGER.warning("Thread for operation {} is not available".format(operation_id))
+        else:
+            operation_thread.stop()
+            while not operation_thread.stopped():
+                LOGGER.info("Thread for operation {} is stopping".format(operation_id))
+        BurstService().persist_operation_state(operation, STATUS_CANCELED)
+        return True
```

Comment:
This is more a question that it is a suggestion:
- Could this Client be simplified and all the Operation/Simulation/H5 related code be pushed towards service/services? I would see here the pyunicore/transport calls left, as well the  HPC very specific parts, while the rest, which are dealing with tvb internal logic could go in other layers ?

Reply:
I have moved most parts that belong to other services or under h5.